### ISSUE TITLE
Support tableprefix

### ISSUE THREAD

**Issue by trondhindenes** [pynamodb/PynamoDB#487]
Since DynamoDB tables are globally unique, it's a bit cumbersome to store the table name inside the model class. It would be awesome to be able to set a "table prefix" on the connection object or similar, so that the model classes themselves could be static and not require dynamic configuration injected. This would allow scenarios like dev/test inside the same aws account etc.

**Comment by numberoverzero** [comment]
I've also found a function (in addition to a prefix or template string) is very useful for automated testing, where I'd use a nonce to isolate integration tests:

```python
nonce = lambda model: model.Meta.table_name + str(time.time())
```

I think this needs to be somewhere on the `Connection` object, looking through https://pynamodb.readthedocs.io/en/latest/low_level.html#creating-a-connection

**Comment by m-melis** [comment]
This would be so useful

**Comment by sirianni** [comment]
> Were tables names globally unique in 2018? They certainly don't seem to be now

I think the original poster meant unique within an AWS account/region.  Given that an organization may have several applications using DynamoDB sharing a single AWS account it would be useful to support a convention where the table names are prefixed with the application name for example (e.g. `app1.Users`).

For example: https://github.com/awspring/spring-cloud-aws/issues/725

**Comment by ikonst** [comment]
At Lyft we got away with initializing `table_name` at load time, i.e.

```py
class MyModel(Model):
     class Meta:
        table_name = env + "_my_model"
```

The environment is normally static per the entire run. And nowadays AWS accounts are often used for env segregation so I'm not even sure that's a very needed pattern.

Of course we can add a hook somewhere higher up to mutate table name (add prefix? add suffix?) but that encourages code where its harder to tell what table name will be used in runtime.

**Comment by sirianni** [comment]
> And nowadays AWS accounts are often used for env segregation so I'm not even sure that's a very needed pattern.

Yes agreed.  This would be more for cases where multiple applications are sharing a single AWS Account (and potentially even accessing each others tables) but you want some loose namespacing or ownership boundaries.

**Comment by ikonst** [comment]
IAM roles can already enforce such a boundary (with `"Resource": "arn:aws:dynamodb:*:*:table/AppName-*"`).

PynamoDB could put in place a convention, but then when you read this code
```python
class Users(Model):
  class Meta:
    table_name = 'users'
```
previously you knew that a table named "users" is what's being used, and now you'd need to cross-check that with configuration.

An alternative is to clarify, right there and then, what's actually being done:
```python
from my_org_lib import make_table_name

class Users(Model):
  class Meta:
    table_name = make_table_name(env, app_name, 'users')
```
In my opinion, this has better Locality of Behavior than the proposal.

**Comment by sirianni** [comment]
Fair point!
